src/validation.cpp

```diff
@@ -2826,14 +2784,13 @@ bool CChainState::ActivateBestChain(BlockValidationState &state, const CChainPar
         LimitValidationInterfaceQueue();
 
         {
-            LOCK2(cs_main, ::mempool.cs); // Lock transaction pool for at least as long as it takes for connectTrace to be consumed
+            // Do not unlock cs_main or mempool until we've made forward
+            // progress (with the exception of shutdown due to hardware issues,
```

Comment:
nit: define forward progress (chain tip is the valid most-work chain or most-work chain has been found invalid)